### PR TITLE
feature: add corresponding Lua API to ssl client hello cb hook

### PR DESCRIPTION
**Pull request by catbro666**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

see [lua-nginx-module](https://github.com/openresty/lua-nginx-module/pull/1900) for details

### CODE REVIEW

File: lib/ngx/ssl/clienthello.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+[Back to TOC](#table-of-contents)
+
+get_client_hello_ext
+----------------------
+**syntax:** *ext, err = ssl_clt.get_client_hello_ext(ext_type)*
+
+**context:** *ssl_client_hello_by_lua&#42;*
+
+Returns raw data of arbitrary SSL client hello extension including custom extensions.
```
[the remark this answers is not in the document]
Comment (catbro666) [reply lib/ngx/ssl/clienthello.md:115]: Could you please have a look at the CI test failure, I could not figure out the reason of this error `Can't connect to 127.0.0.1:25382: Connection refused`

[the remark this answers is not in the document]
Comment (spacewander) [reply lib/ngx/ssl/clienthello.md:115]: Look like a crash of the OpenResty.

**Review by zhuizhuhaomeng — APPROVED** [review]

**Comment by zhuizhuhaomeng** [comment]
```
diff --git a/lib/ngx/ssl/clienthello.lua b/lib/ngx/ssl/clienthello.lua
index 10cae80..ed68f9b 100644
--- a/lib/ngx/ssl/clienthello.lua
+++ b/lib/ngx/ssl/clienthello.lua
@@ -15,6 +15,7 @@ local errmsg = base.get_errmsg_ptr()
 local get_size_ptr = base.get_size_ptr
 local FFI_OK = base.FFI_OK
 local subsystem = ngx.config.subsystem
+local ngx_phase = ngx.get_phase
 
 
 local ngx_lua_ffi_ssl_get_client_hello_server_name
@@ -77,6 +78,10 @@ function _M.get_client_hello_server_name()
         error("no request found")
     end
 
+    if ngx_phase() ~= "ssl_client_hello" then
+        error("API disabled in the current context", 2)
+    end
+
     local sizep = get_size_ptr()
 
     local rc = ngx_lua_ffi_ssl_get_client_hello_server_name(r, ccharpp, sizep,
@@ -85,6 +90,11 @@ function _M.get_client_hello_server_name()
         return ffi_str(ccharpp[0], sizep[0])
     end
 
+    -- NGX_DECLINED: no sni extension
+    if rc == -5 then
+        return nil
+    end
+
     return nil, ffi_str(errmsg[0])
 end
 
@@ -96,14 +106,23 @@ function _M.get_client_hello_ext(ext_type)
         error("no request found")
     end
 
+    if ngx_phase() ~= "ssl_client_hello" then
+        error("API disabled in the current context", 2)
+    end
+
     local sizep = get_size_ptr()
 
     local rc = ngx_lua_ffi_ssl_get_client_hello_ext(r, ext_type, cucharpp,
-                sizep, errmsg)
+                                                    sizep, errmsg)
     if rc == FFI_OK then
         return ffi_str(cucharpp[0], sizep[0])
     end
 
+    -- NGX_DECLINED: no extension
+    if rc == -5 then
+        return nil
+    end
+
     return nil, ffi_str(errmsg[0])
 end
 
@@ -125,6 +144,10 @@ function _M.set_protocols(protocols)
         error("no request found")
     end
 
+    if ngx_phase() ~= "ssl_client_hello" then
+        error("API disabled in the current context" .. ngx_phase(), 2)
+    end
+
     local prots = 0
     for _, v in ipairs(protocols) do
         if not prot_map[v] then
diff --git a/lib/ngx/ssl/clienthello.md b/lib/ngx/ssl/clienthello.md
index 3c59669..bc70b4c 100644
--- a/lib/ngx/ssl/clienthello.md
+++ b/lib/ngx/ssl/clienthello.md
@@ -67,7 +67,7 @@ server {
 Description
 ===========
 
-This Lua module provides API functions for post-processing SSL client hello message for NGINX downstream connections. 
+This Lua module provides API functions for post-processing SSL client hello message for NGINX downstream connections.
 
 It must to be used in the contexts [ssl_client_hello_by_lua*](https://github.com/openresty/lua-nginx-module/#ssl_client_hello_by_lua_block).
 
@@ -75,7 +75,7 @@ This Lua API is particularly useful for dynamically setting the SSL protocols ac
 
 It is also useful to do some custom operations according to the per-connection information in the client hello message.
 
-For example, one can parse custom client hello extension and do the corresponding handling in pure Lua.
+For example, one can parse the custom client hello extension and do the corresponding handling in pure Lua.
 
 To load the `ngx.ssl.clienthello` module in Lua, just write
 
@@ -94,13 +94,13 @@ get_client_hello_server_name
 
 **context:** *ssl_client_hello_by_lua&#42;*
 
-Returns the TLS SNI (Server Name Indication) name set by the client. 
+Returns the TLS SNI (Server Name Indication) name set by the client.
 
 Return `nil` when then the extension does not exist.
 
 In case of errors, it returns `nil` and a string describing the error.
 
-Note that the SNI name is gotten from the raw extensions of client hello message associated with the current downstream SSL connection.
+Note that the SNI name is gotten from the raw extensions of the client hello message associated with the current downstream SSL connection.
 
 So this function can only be called in the context of [ssl_client_hello_by_lua*](https://github.com/openresty/lua-nginx-module/#ssl_client_hello_by_lua_block).
 
@@ -112,13 +112,13 @@ get_client_hello_ext
 
 **context:** *ssl_client_hello_by_lua&#42;*
 
-Returns raw data of arbitrary SSL client hello extension including custom extensions. 
+Returns raw data of arbitrary SSL client hello extension including custom extensions.
 
 Returns `nil` if the specified extension type does not exist.
 
 In case of errors, it returns `nil` and a string describing the error.
 
-Note that the ext is gotten from the raw extensions of client hello message associated with the current downstream SSL connection.
+Note that the ext is gotten from the raw extensions of the client hello message associated with the current downstream SSL connection.
 
 So this function can only be called in the context of [ssl_client_hello_by_lua*](https://github.com/openresty/lua-nginx-module/#ssl_client_hello_by_lua_block).
 
@@ -179,11 +179,11 @@ set_protocols
 
 **context:** *ssl_client_hello_by_lua&#42;*
 
-Sets the SSL protocols supported by the current downstream SSL connection. 
+Sets the SSL protocols supported by the current downstream SSL connection.
 
-Returns `true` on success, or a `nil` value and a string describing the error otherwise. 
+Returns `true` on success, or a `nil` value and a string describing the error otherwise.
 
-Considering it is meaningless to set ssl protocols after the protocol is determined, 
+Considering it is meaningless to set ssl protocols after the protocol is determined,
 so this function may only be called in the context of [ssl_client_hello_by_lua*](https://github.com/openresty/lua-nginx-module/#ssl_client_hello_by_lua_block).
 
 Example: `ssl_clt.set_protocols({"TLSv1.1", "TLSv1.2", "TLSv1.3"})`
diff --git a/t/ssl-client-hello.t b/t/ssl-client-hello.t
index dc3c649..7dcf9e7 100644
--- a/t/ssl-client-hello.t
+++ b/t/ssl-client-hello.t
@@ -374,6 +374,7 @@ read SNI name from Lua: test.com
             local ext = ssl_clt.get_client_hello_ext(0)
             print("read SNI name from Lua: ", ext, ", type: ", type(ext))
         }
+
         ssl_certificate ../../cert/test.crt;
         ssl_certificate_key ../../cert/test.key;
 
```
@catbro666  please apply the patch, so others can review it without asking the same question?

**Comment by zhuizhuhaomeng** [comment]
to make the ci pass, you need to rebase the openresty-devel-utils that you forked.

**Comment by catbro666** [comment]
@zhuizhuhaomeng Thanks, the patch is applied and the repos are updated.

**Review by doujiang24 — APPROVED** [review]

**Comment by zhuizhuhaomeng** [comment]
merged.
Thank you for your contribution.

**Pull request closed without merging**